src/index.ts

```diff
@@ -0,0 +1,236 @@
+} from "./message";
+import { TALKER_TYPE, TALKER_ERR_TIMEOUT } from "./strings";
+
+interface SentMessages {
+  [id: number]: ManipulablePromise<IncomingMessage | Error>;
+}
+
+/**
+ * Talker
+ * Opens a communication line between this window and a remote window via postMessage.
+ */
+class Talker {
+  /*
+   * @property timeout - The number of milliseconds to wait before assuming no response will be received.
+   */
+  public timeout: number = 3000;
+
+  /**
+   * @property onMessage - Will be called with every non-handshake, non-response message from the remote window
+   */
+  onMessage?: (message: IncomingMessage) => void;
+
+  // Will be resolved when a handshake is newly established with the remote window.
+  private readonly handshake: ManipulablePromise<boolean>;
+  // Whether we've received a handshake from the remote window
+  private handshaken: boolean;
+  // The ID of the latest OutgoingMessage
+  private latestId: number = 0;
+  private readonly queue: OutgoingMessage[] = [];
+  private readonly sent: SentMessages = {};
+
+  /**
+   * @param remoteWindow - The remote `window` object to post/receive messages to/from
+   * @param remoteOrigin - The protocol, host, and port you expect the remoteWindow to be
+   * @param localWindow - The local `window` object
+   */
+  constructor(
+    private readonly remoteWindow: Window,
+    private readonly remoteOrigin: string,
+    private readonly localWindow: Window = window
+  ) {
+    this.handshaken = false;
+    this.handshake = createManipulablePromise();
+
+    this.localWindow.addEventListener(
+      "message",
+      (messageEvent: MessageEvent) => this.receiveMessage(messageEvent),
+      false
+    );
+    this.sendHandshake();
+
+    return this;
+  }
+
+  /**
+   * @param namespace - The namespace the message is in
+   * @param data - The data to send
+   * @param responseToId - If this is a response to a previous message, its ID.
+   */
+  send(
+    namespace: string,
+    data: Stringifyable,
+    responseToId: number | null = null
+  ): ManipulablePromise<IncomingMessage | Error> {
+    const message: OutgoingMessage = new OutgoingMessage(
+      this,
+      namespace,
+      data,
+      responseToId
+    );
+
+    const promise = createManipulablePromise<IncomingMessage | Error>();
+
+    this.sent[message.id] = promise;
+    this.queue.push(message);
+    this.flushQueue();
+
+    setTimeout(
+      () =>
+        promise.__reject__ && promise.__reject__(new Error(TALKER_ERR_TIMEOUT)),
+      this.timeout
+    );
+
+    return promise;
+  }
+
+  /**
+   * This is not marked private because other Talker-related classes need access to it,
+   * but your application code should probably avoid calling this method.
+   */
+  nextId(): number {
+    return (this.latestId += 1);
+  }
+
+  private receiveMessage(messageEvent: MessageEvent): void {
+    let object: JSONifiedMessage;
+    try {
+      object = JSON.parse(messageEvent.data);
+    } catch (err) {
+      object = {
```

Comment:
I'm not that familiar with talker; can you explain why we'd ever want to keep going here if the message data failed JSON parsing? (And preferably add a comment to the code.)

Reply:
It's a feature of the library that data-less messages can be sent via Talker. This handles that scenario (e.g. `JSON.parse(undefined)`, `JSON.parse('')`, etc.).